### ISSUE TITLE
docs-bug(*): Missing migration guide to v19

### ISSUE THREAD

**Issue by martinboue** [angular/components#30055]
### Documentation Feedback

I'm struggling to migrate from Angular Material v18 (with MD3) to v19. I did not find any migration guide and the new guides do not explain the basics as deeply as in the previous version.

Here are the most important use cases that I don't know how to deal with:

### Before
```
.my-component {
   background-color: mat.get-theme-color(theme.$theme, neutral, 98)
}
```
### After
I cannot use mat.get-theme-color because I don't have a $theme anymore (previously from mat.define-theme). How can I do that?

### Before:
I had either this:
```
:root {
   @include mat.all-component-themes(theme.$theme);
}
```
or this:

```
:root {
   @include mat.button-theme(theme.$theme);
   @include mat.sidenav-theme(theme.$theme);
   // ...
}
```

### After
Is it still required? I'm guessing it's not and I can remove everything but I'm not sure.

### Before
```
@include mat.core();
:root {
   @include mat.core-theme(theme.$light-theme);
}
```

### After 
I has been automatically replaced to:

```
@include mat.elevation-classes();
@include mat.app-background();
:root {
   @include mat.elevation-classes();
   @include mat.app-background();
}
```
What is it for? Why isn't it present if I create a new application with Angular Material? Should it be duplicated? I don't see any documentation mentioning those mixins.

### Before
I could do this:
```
@include mat.chips-color(theme.$theme, $color-variant: primary);
@include mat.form-field-density(-3);
```

### After
How to do the same? Do I need to use the new overrides mixins and set each color tokens? For example, that would be 25 colors for the chip component.
 

### Affected documentation page

https://material.angular.io/guide/theming

**Comment by X-Celcius** [comment]
This is not only an issue when you want to migrate to the new `mat.theme` and override mixins. According to the comment in `_core.scss`, `mat.core()` is deprecated and doesn't do anything in v19. It took us some time to find out why `.mat-app-background` was no longer being defined when updating to v19. 
If this is a breaking change it should at least be mentioned in the Changelog, but I can't find `mat.core` being mentioned there either.

**Comment by X-Celcius** [comment]
And while this is not related to documentation, may I also suggest to keep the scss in `mat.core` as it was and just add the deprecation annotation / warning rather than immediately changing the behavior and adding a comment that it is going to be removed in v21? Because if you already have to modify your code to get the old behavior why keep an empty mixin around?

**Comment by andrewseguin** [comment]
Thanks for the feedback. Our previous theme config system was fairly complex and required a deep guide on its mechanics. We're hoping this new system is more straightforward with how it operates, but this means the guide might feel more lightweight than before.

### Styling custom components

In the same pattern of system CSS variables, I'd recommend simply creating a top-level CSS variable for colors you want to use within your application. E.g. 

```
--my-app-theme-neutral98: map.get($my-palette, neutral, 98);
```

This bypasses the need to create a separate mixin in your component styles that accepts a theme config, and allows you to just put the style directly in your component style.

If you want to continue using the mixin concept, you can pass it the palette or color directly. 

Otherwise, you can always continue creating a theme config that can be passed around, in parallel to using `mat.theme`.

### Material 3 theme setup

`mat.core` was replaced by `elevation-classes` and `app-background` since those styles were previously included in its old implementation. However, `mat.core-theme` should not have been replaced. If that happened in the update schematic, it should be filed as a bug.

For a new app, we don't really recommend using `elevation-classes` or `app-background` anymore. Elevation can be applied using the system level variables instead. And the app background can be set yourself if necessary (this is pointed out in the theming guide), which has the benefit of being explicit rather than a black-box mixin.

### Customizing material components

What are you trying to achieve in the line for chips-color? Isn't the default going to be primary? 

If you are trying to set it with a different color, you can write a selector that targets your mat-chips and redefine the theme as described in https://material.angular.io/guide/theming#context-specific-themes

This applies for density as well - you can target the form field and call `mat.theme((density: -3));` or otherwise call the form field's `overrides` mixin and directly customize the tokens, e.g. 

```
  @include mat.form-field-overrides((
    container-height: 44px,
    container-vertical-padding: 10px,
    filled-with-label-container-padding-top: 10px,
    filled-with-label-container-padding-bottom: 10px,
  ));
```

However, the `density` mixin still exists and you can continue calling `mat.form-field-density(-3)`

**Comment by andrewseguin** [comment]
> This is not only an issue when you want to migrate to the new `mat.theme` and override mixins. According to the comment in `_core.scss`, `mat.core()` is deprecated and doesn't do anything in v19. It took us some time to find out why `.mat-app-background` was no longer being defined when updating to v19. If this is a breaking change it should at least be mentioned in the Changelog, but I can't find `mat.core` being mentioned there either.

`mat.core` should have been replaced through the schematic to be:

```
@include mat.elevation-classes();
@include mat.app-background();
```

If this wasn't the case, it'd be helpful to have an issue filed. Our intention is to keep the same behavior/styling preserved for applications in the update process

**Comment by rars** [comment]
It's not obvious to me how to handle typography for non-component styles. The default styles of `h1`, `h2` etc. don't seem to look very nice with just the `@include mat.theme(...)` approach (I think the mixin doesn't touch the styles on these tags). I used to include the `typography-hierarchy` mixin and use the classes listed in [type scale levels](https://v18.material.angular.io/guide/typography#type-scale-levels). I couldn't find relevant documentation and if it's missing, I think the documentation would benefit from touching on this subject.

**Comment by X-Celcius** [comment]
> If this wasn't the case, it'd be helpful to have an issue filed. Our intention is to keep the same behavior/styling preserved for applications in the update process

Thanks, I wasn't aware of an update schematic. My colleague updated the Angular core libraries using `ng update` before the cdk & materials packages were available in v19. A day later, he updated the material packages directly without running `ng update`. I retried going v18 to v19 from scratch and everything updated automatically so this is actually working fine. 
Still - please consider adding this change to [CHANGELOG.md](https://github.com/angular/components/blob/main/CHANGELOG.md) as I consider this to be a breaking change and it is not mentioned there. Automatic migrations are nice if they work but knowing what and why something will change is also essential when upgrading.

**Comment by martinboue** [comment]
Because some of my comments on this issue goes beyond the scope "Migrating v18 to v19", I have created a specific request on "how to get theme values without relying on private APIs", see #30083 

My main concern now is: How to _easily_ migrate v18 `mat.<component>-color` mixins to v19?

If the solution is using the new `mat.<component>-overrides` mixin, could you consider adding a new mixin to quickly change color for a single component without specifying every color tokens?

**Comment by martinboue** [comment]
For context, here is a before/after comparison on how to change the chip component color:

Before in v18:
```
@include mat.chips-color(theme.$theme, $color-variant: primary);
```

After in v19:
```
@include mat.chips-overrides((
   elevated-selected-container-color: map.get(theme.$palettes, primary, 90),
   selected-label-text-color: map.get(theme.$palettes, primary, 10),
   selected-trailing-icon-color: map.get(theme.$palettes, primary, 10),
   selected-hover-state-layer-color: map.get(theme.$palettes, primary, 10),
   selected-trailing-action-state-layer-color: map.get(theme.$palettes, primary, 10),
   selected-focus-state-layer-color: map.get(theme.$palettes, primary, 10),
   with-icon-selected-icon-color: map.get(theme.$palettes, primary, 10)
));
```
It's much more complicated now, and if we want to keep the same level of contrast (here tones 10 and 90) we have to inspect the generated code and find the equivalent in the new palette.

### Proposed solution
In v18 we had to pick a color from our theme, which wasn't very flexible. An ideal solution would be:
```
@include mat.chips-color(#FF0000); // <== We could pass any color here and the mixin would compute the tones
```
If not doable, then the same behavior as v18 would be good enough:
```
@include mat.chips-color(theme.$palettes, primary);
// or
$primary-palette: map.get(theme.$palettes, primary);
@include mat.chips-color($primary-palette);
```

**Comment by edvinv** [comment]
For example in 18 you can define:
``` css
.secondary-button {
  @include mat.button-color($theme, $color-variant: secondary);
}
```

In 19 you have :
``` css
.secondary-button {
   @include mat.button-overrides((
    filled-container-color: orange,
    filled-label-text-color: red,
    ....
  ));
}
```

For some control like icon, it is easy because there is single color, but for other it may be quite complicated. For now, as a workaround I have just done:

``` css
.secondary-button {
    --mat-sys-primary: var(--mat-sys-secondary);
  }
```
This works but it is just temporally solution. IMO mat.{component}-overrides should support changing color with single input parameter (for controls that are based on single color) like:

``` css
.secondary-button {
   @include mat.button-overrides((
    primary-color: var(--mat-sys-secondary),
  ));
}
```

**Comment by andrewseguin** [comment]
> It would be great to have public APIs to rely on instead of CSS variables that cannot be checked during compilation."

@martinboue The system level variables should be considered part of the public API. Any changes made to them in the future will be considered breaking and I'd expect us to build schematics to auto-update folks (e.g. renaming or removing variables)

> Is there any way to generate only component related properties?

@martinboue You can use the component mixins like `mat.chip-density(-3)` to output only chip-related density tokens

> Also, now I need to specify every kind of chips (mat-chip, mat-chip-row and mat-chip-option) and rely on private CSS class or material host tag, both of which can change in the future and are not checked at compile time. Is there any other solution?

@martinboue You shouldn't need to target a specific selector, just include that in your `html` or wherever you want that CSS variable definition to apply

> It's not obvious to me how to handle typography for non-component styles.

@rars The typography hierarchy was a Material 2 typography spec, but that is no longer defined for Material 3. We did our best to mimic a reasonable hierarchy, but since it's not part of the spec I think it makes more sense for clients to set it themselves. You can use this as a reference and modify it if needed:

```css
h1 {
    font: var(--mat-sys-display-large);
}

h2 {
    font: var(--mat-sys-display-medium);
}

h3 {
    font: var(--mat-sys-display-small);
}

h4 {
    font: var(--mat-sys-headline-large);
}

h5 {
    font: var(--mat-sys-headline-medium);
}

h6 {
    font: var(--mat-sys-headline-small);
}
```

**Comment by martinboue** [comment]
@andrewseguin

> The system level variables should be considered part of the public API. Any changes made to them in the future will be considered breaking and I'd expect us to build schematics to auto-update folks (e.g. renaming or removing variables)

From my understanding, system variables are useful for changing global styles or reusing it for non-material components but are not ideal for customizing material components.

> You can use the component mixins like mat.chip-density(-3) to output only chip-related density tokens

Yes it works for density, but what about color? It's a very common use case to change a material component color but there is no way to do it easily like we used to with `mat.<component>-color` mixin.

Could you please give us an example in v19 of how to change the color of a chip where the result is the equivalent of the `mat.chip-color` mixin in v18?

**Comment by yusufakg** [comment]
Expanding on @andrewseguin input. For anyone that is updating to the new mat.theme and used typography-hierachy before here is what I did to get the same effect.

theme-typography.scss:
```
@mixin my-typography-hierarchy {
  // Display styles
  h1 {
    font: var(--mat-sys-display-large);
    letter-spacing: var(--mat-sys-display-large-tracking);
    margin: 0 0 16px;
  }

  h2 {
    font: var(--mat-sys-display-medium);
    letter-spacing: var(--mat-sys-display-medium-tracking);
    margin: 0 0 16px;
  }

  h3 {
    font: var(--mat-sys-display-small);
    letter-spacing: var(--mat-sys-display-small-tracking);
    margin: 0 0 16px;
  }

  // Headline styles
  h4 {
    font: var(--mat-sys-headline-large);
    letter-spacing: var(--mat-sys-headline-large-tracking);
    margin: 0 0 16px;
  }

  h5 {
    font: var(--mat-sys-headline-medium);
    letter-spacing: var(--mat-sys-headline-medium-tracking);
    margin: 0 0 12px;
  }

  h6 {
    font: var(--mat-sys-headline-small);
    letter-spacing: var(--mat-sys-headline-small-tracking);
    margin: 0 0 12px;
  }

  // Body styles
  body {
    font: var(--mat-sys-body-large);
    letter-spacing: var(--mat-sys-body-large-tracking);

    p {
      margin: 0 0 12px;
    }
  }

  // Additional body classes
  .body-small {
    font: var(--mat-sys-body-small);
    letter-spacing: var(--mat-sys-body-small-tracking);
  }

  .body-medium {
    font: var(--mat-sys-body-medium);
    letter-spacing: var(--mat-sys-body-medium-tracking);
  }

  .body-large {
    font: var(--mat-sys-body-large);
    letter-spacing: var(--mat-sys-body-large-tracking);
  }

  // Additional display classes
  .display-small {
    font: var(--mat-sys-display-small);
    letter-spacing: var(--mat-sys-display-small-tracking);
  }

  .display-medium {
    font: var(--mat-sys-display-medium);
    letter-spacing: var(--mat-sys-display-medium-tracking);
  }

  .display-large {
    font: var(--mat-sys-display-large);
    letter-spacing: var(--mat-sys-display-large-tracking);
  }

  // Additional headline classes
  .headline-small {
    font: var(--mat-sys-headline-small);
    letter-spacing: var(--mat-sys-headline-small-tracking);
  }

  .headline-medium {
    font: var(--mat-sys-headline-medium);
    letter-spacing: var(--mat-sys-headline-medium-tracking);
  }

  .headline-large {
    font: var(--mat-sys-headline-large);
    letter-spacing: var(--mat-sys-headline-large-tracking);
  }

  // Additional label classes
  .label-small {
    font: var(--mat-sys-label-small);
    letter-spacing: var(--mat-sys-label-small-tracking);
  }

  .label-medium {
    font: var(--mat-sys-label-medium);
    letter-spacing: var(--mat-sys-label-medium-tracking);
  }

  .label-large {
    font: var(--mat-sys-label-large);
    letter-spacing: var(--mat-sys-label-large-tracking);
  }

  // Additional title classes
  .title-small {
    font: var(--mat-sys-title-small);
    letter-spacing: var(--mat-sys-title-small-tracking);
  }

  .title-medium {
    font: var(--mat-sys-title-medium);
    letter-spacing: var(--mat-sys-title-medium-tracking);
  }

  .title-large {
    font: var(--mat-sys-title-large);
    letter-spacing: var(--mat-sys-title-large-tracking);
  }
}
```
And then i use it in styles.scss:
```
...
@use "./theme-typography.scss" as m3-theme-typography;

html {
  @include mat.theme((
    ...
  ));

  @include m3-theme-typography.my-typography-hierarchy;
}
```

I do find this slightly better mainly because of future discoverability/maintainability and while I understand the reasoning, I just wished it was mentioned somewhere without searching through issues.

**Comment by floinay** [comment]
Recently, I spent several days figuring out how to style a theme in Material 18, including working with use-system-variables, system-variables-prefix, density, and typography, because the documentation is very superficial.

Now, I need to do all of this again with Material 19. Additionally, the old API is not fully supported, but the code does not throw any errors, even though my theme no longer works as expected after migration.

You really need to improve the documentation and include more examples with links to various templates on GitHub. Perhaps it’s worth considering hiring someone specifically to handle this aspect.

**Comment by JoepKockelkorn** [comment]
I've recently investigated how to use Material 18 and thought I had it all figured out. Now, when actually starting the project and using Material 19 I find out the typography mixin is now missing. When/where was that communicated? I inspected the [CHANGELOG](https://github.com/angular/components/blob/main/CHANGELOG.md) but can't seem to find the moment in time or a related reference.

**Comment by martinboue** [comment]
As there are several subjects in this issue, I have a created a specific feature request on "How to change a material component color" which sums up the original problem I had, here: https://github.com/angular/components/issues/30241

**Comment by roma2341** [comment]
Guys, what means "overrides API" ?

**Comment by marc-wilson** [comment]
Why does every single release of angular since 15 have some crazy overhaul to the theming system? It's very mysterious and poorly documented. Just like @JoepKockelkorn, I thought I had everything figured out and working in 18. Just upgraded to 19 and now typography and coloring is broken.

After the disastrous MDC migration and a new theming system with every version, it's hard to pitch angular as a stable framework to use when we are constantly chasing our tail trying keep up with angular.

Please stabilize theming as advertised in 18.

**Comment by roma2341** [comment]
How to restyle button when it hovered ?
Now i use this, but i'm not sure that it's correct:
```:root {
  @include mat.button-overrides(
    (
      outlined-label-text-color: var(--colors-elements-main),
      filled-container-color: var(--colors-surfaces-main),
      filled-container-shape: var(--radiuses-s),
      filled-label-text-color: var(--colors-elements-on-main),
      filled-label-text-font: var(--crm-label-m-font-family),
      filled-label-text-weight: var(--crm-label-m-font-weight),
      filled-disabled-container-color: var(--colors-main-states-disabled),
      /*filled-hover-state-layer-opacity not working*/
    )
  );
}
button:hover {
  @include mat.button-overrides(
    (
      filled-container-color: red,
    )
  );
}```
